core/unix/coredump.c

```diff
@@ -0,0 +1,360 @@
+    }
+    // Write memory content to the core dump file.
+    for (int section_index = 0; section_index < section_count - 1; ++section_index) {
+        const size_t length = section_header_info[section_index].vm_end -
+            section_header_info[section_index].vm_start;
+        const size_t written = os_write(
+            elf_file, (void *)section_header_info[section_index].vm_start, length);
+        if (written != length) {
+            SYSLOG_INTERNAL_ERROR("Failed to write the requested memory content into "
+                                  "the core dump file.");
+            os_close(elf_file);
+            return false;
+        }
+    }
+    // Write the section names section.
+    if (os_write(elf_file, (void *)string_table, string_table_offset) !=
+        string_table_offset) {
+        SYSLOG_INTERNAL_ERROR("Failed to write section name string table into the "
+                              "core dump file.");
+        os_close(elf_file);
+        return false;
+    }
+    // Write section headers to the core dump file.
+    // TODO i#7046: Handle multiple program headers.
+    ELF_OFF file_offset = sizeof(ELF_HEADER_TYPE) +
+        1 /*program_header_count*/ * sizeof(ELF_PROGRAM_HEADER_TYPE);
+    for (int section_index = 0; section_index < section_count - 1; ++section_index) {
+        ELF_WORD flags = SHF_ALLOC | SHF_MERGE;
+        if (section_header_info[section_index].prot & PROT_WRITE) {
+            flags |= SHF_WRITE;
+        }
+        if (!write_section_header(
+                elf_file, section_header_info[section_index].name_offset, SHT_PROGBITS,
+                flags, (ELF_ADDR)section_header_info[section_index].vm_start, file_offset,
+                section_header_info[section_index].vm_end -
+                    section_header_info[section_index].vm_start,
+                /*link=*/0,
+                /*info=*/0, /*alignment=*/sizeof(ELF_WORD),
+                /*entry_size=*/0)) {
+            os_close(elf_file);
+            return false;
+        }
+        file_offset += section_header_info[section_index].vm_end -
+            section_header_info[section_index].vm_start;
+    }
+    // Write the section names section.
+    if (!write_section_header(
+            elf_file, string_table_offset - strlen(SECTION_HEADER_TABLE), SHT_STRTAB,
+            /*flags=*/0, /*virtual_address=*/0, file_offset,
+            /*section_size=*/string_table_offset, /*link=*/0,
+            /*info=*/0, /*alignment=*/1,
+            /*entry_size=*/0)) {
+        os_close(elf_file);
+        return false;
+    }
+
+    os_close(elf_file);
+    strhash_hash_destroy(GLOBAL_DCONTEXT, string_htable);
+    return true;
+}
+
+/*
+ * Returns true if a core dump file is written, false otherwise.
+ */
+bool
+os_dump_core_live(void)
+{
+    static thread_id_t current_dumping_thread_id VAR_IN_SECTION(NEVER_PROTECTED_SECTION) =
+        0;
+    thread_id_t current_id = d_r_get_thread_id();
+#ifdef DEADLOCK_AVOIDANCE
```

Comment:
I think this should just be removed.

Reply:
I've removed all the DEADLOCK_AVOIDANCE code.